cpp/include/cudf/io/parquet_metadata.hpp

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file parquet_metadata.hpp
+ * @brief cuDF-IO freeform API
+ */
+
+#pragma once
+
+#include <cudf/io/types.hpp>
+
+#include <optional>
+#include <string_view>
+#include <variant>
+#include <vector>
+
+namespace cudf {
+namespace io {
+
+namespace parquet {
+/**
+ * @brief Basic data types in Parquet, determines how data is physically stored
+ */
+enum class TypeKind : int8_t {
+  UNDEFINED_TYPE       = -1,  // Undefined for non-leaf nodes
+  BOOLEAN              = 0,
+  INT32                = 1,
+  INT64                = 2,
+  INT96                = 3,  // Deprecated
+  FLOAT                = 4,
+  DOUBLE               = 5,
+  BYTE_ARRAY           = 6,
+  FIXED_LEN_BYTE_ARRAY = 7,
+};
+}  // namespace parquet
+
+/**
+ * @brief Schema of a parquet column, including the nested columns.
+ */
+struct parquet_column_schema {
+ public:
+  /**
+   * @brief constructor
+   *
+   * @param name column name
+   * @param type parquet type
+   * @param children child columns (empty for non-nested types)
+   */
+  parquet_column_schema(std::string_view name,
+                        parquet::TypeKind type,
+                        std::vector<parquet_column_schema> children)
+    : _name{name}, _type_kind{type}, _children{std::move(children)}
+  {
+  }
+
+  /**
+   * @brief Returns parquet column name; can be empty.
+   *
+   * @return Column name
+   */
+  [[nodiscard]] auto name() const { return _name; }
+
+  /**
+   * @brief Returns parquet type of the column.
+   *
+   * @return Column parquet type
+   */
+  [[nodiscard]] auto type_kind() const { return _type_kind; }
+
+  /**
+   * @brief Returns schemas of all child columns.
+   *
+   * @return Children schemas
+   */
+  [[nodiscard]] auto const& children() const& { return _children; }
+
+  /** @copydoc children
+   * Children array is moved out of the object (rvalues only).
+   *
+   */
+  [[nodiscard]] auto children() && { return std::move(_children); }
```

Comment:
What is the necessity of this?

Reply:
@vuule 's suggestion here https://github.com/rapidsai/cudf/pull/13663#discussion_r1270909433

Reply:
For rvalues objects, `.children()` can return rvalue. It will be useful at places where only this member of rvalue objects are accessed.

Reply:
Oh okay, if you expect the pattern to be of the kind `parquet_column_schema().children()` then it makes sense to have these overloads